src/services/BaseAPIServices.ts

```diff
@@ -0,0 +1,467 @@
+import axios, { AxiosError, ResponseType } from 'axios';
+import { authenticateSilently } from './auth';
+import { metricKeys, metricStatus, track } from './appInsights';
+import * as FileSystem from 'expo-file-system';
+import * as Device from 'expo-device';
+
+export type BaseResource = {
+  scopes: string[];
+  apiBaseUrl: string;
+  subscriptionKey?: string;
+};
+
+export type BaseAPIOptions = {
+  authenticate?: boolean;
+  headers?: Record<string, any>;
+  responseType?: ResponseType;
+};
+
+export type DownloadFileOptions = BaseAPIOptions & {
+  filePath?: string;
+};
+
+const defaultOptions = {
+  authenticate: true,
+};
+
+const defaultFileDownloadOptions = {
+  ...defaultOptions,
+  filePath: FileSystem.cacheDirectory,
+};
+
+class BaseApiService {
+  url: string;
+
+  scopes: string[];
+
+  subscriptionKey?: string;
+
+  appSpecificDefaultHeaderFunction?: (() => Record<string, any>) | (() => {});
```

Comment:
Is there a reason this is defined as a function and not an object ?

Reply:
Great question actually!
It's a function because we want the output to be calculated on each request instead of on app start. Things like app language can change during the use of the app, but if `appSpecificDefaultHeaderFunction` was an already calculated object, the user would have to kill the app and restart it to switch the language of the response of api calls.